File: Bots/DotNet/Consumers/CodeFirst/WaterfallHostBot/Dialogs/Sso/SsoDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Globalization;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Builder.Integration.AspNet.Core.Skills;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallHostBot.Dialogs.Sso
{
    public class SsoDialog : ComponentDialog
    {
        public SsoDialog(ConversationState conversationState, SkillsConfiguration skillsConfig, SkillHttpClient skillClient, IConfiguration configuration, SkillConversationIdFactoryBase conversationIdFactory)
            : base(nameof(SsoDialog))
        {
            AddDialog(new SignInDialog(configuration));
            AddDialog(new SignOutDialog(configuration));
            AddDialog(new DisplayTokenDialog(configuration));

            var botId = configuration.GetSection("MicrosoftAppId")?.Value;

            AddDialog(new ChoicePrompt("ActionStepPrompt"));
            AddDialog(new ChoicePrompt("FinalStepPrompt"));

            skillsConfig.Skills.TryGetValue("WaterfallSkillBot", out var skill);
            AddDialog(new SkillDialog(
                new SkillDialogOptions()
                {
                    BotId = botId,
                    ConversationIdFactory = conversationIdFactory,
                    ConversationState = conversationState,
                    Skill = skill,
                    SkillClient = skillClient,
                    SkillHostEndpoint = skillsConfig.SkillHostEndpoint
                },
                nameof(SkillDialog)));

            var waterfallSteps = new WaterfallStep[]
            {
                PromptActionStepAsync,
                HandleActionStepAsync,
                PromptFinalStepAsync,
                HandleFinalStepAsync
            };

            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), waterfallSteps));

            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> PromptActionStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var messageText = "What do you want to do?";
            var repromptMessageText = "That was not a valid choice, please select a valid choice.";
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
                Choices = new List<Choice> { new Choice("Login"), new Choice("Logout"), new Choice("Show token"), new Choice("Call Skill"), new Choice("End") }
            };

            // Prompt the user to select a skill.
            return await stepContext.PromptAsync("ActionStepPrompt", options, cancellationToken);
        }

        private async Task<DialogTurnResult> HandleActionStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var action = ((FoundChoice)stepContext.Result).Value.ToLowerInvariant();

            switch (action)
            {
                case "login":
                    return await stepContext.BeginDialogAsync(nameof(SignInDialog), null, cancellationToken);       

                case "logout":
                    return await stepContext.BeginDialogAsync(nameof(SignOutDialog), null, cancellationToken);

                case "show token":
                    return await stepContext.BeginDialogAsync(nameof(DisplayTokenDialog), null, cancellationToken);

                case "call skill":
                    stepContext.Context.Activity.Type = ActivityTypes.Event;
                    stepContext.Context.Activity.Name = "Sso";
                    return await stepContext.BeginDialogAsync(nameof(SkillDialog), new BeginSkillDialogOptions() { Activity = stepContext.Context.Activity }, cancellationToken);

                default:
                    // This should never be hit since the previous prompt validates the choice
                    return new DialogTurnResult(DialogTurnStatus.Complete);
            }
        }

        private async Task<DialogTurnResult> PromptFinalStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var messageText = "What do you want to do?";
            var repromptMessageText = "That was not a valid choice, please select a valid choice.";
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
                Choices = new List<Choice> { new Choice("Another action"), new Choice("End") }
            };

            // Prompt the user to select a skill.
            return await stepContext.PromptAsync("FinalStepPrompt", options, cancellationToken);
        }

        private async Task<DialogTurnResult> HandleFinalStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var action = ((FoundChoice)stepContext.Result).Value.ToLowerInvariant();

            if (action == "another action")
            {
                return await stepContext.ReplaceDialogAsync(InitialDialogId, "What do you want to do?", cancellationToken);
            }

            return await stepContext.EndDialogAsync(null, cancellationToken);
        }
    }
}

File: Bots/JavaScript/Consumers/CodeFirst/WaterfallHostBot/dialogs/sso/ssoDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ActivityTypes, EndOfConversationCodes, InputHints, MessageFactory } = require('botbuilder');
const { ComponentDialog, ChoicePrompt, ChoiceFactory, DialogTurnStatus,WaterfallDialog } = require('botbuilder-dialogs');
const { SsoSignInDialog } = require('./ssoSignInDialog');

const ACTION_STEP_PROMPT = 'ActionStepPrompt';
const DELIVERY_PROMPT = 'DeliveryModePrompt';
const SSO_DIALOG = 'SsoDialog';
const SSO_SIGNIN_DIALOG = 'SsoSignInDialog'
const WATERFALL_DIALOG = 'WaterfallDialog';

// Helps prepare the host for SSO operations and provides helpers to check the status and invoke the skill.
class SsoDialog extends ComponentDialog {
    constructor(skillDialog) {
        super(SSO_DIALOG + skillDialog.id);

        this.connectionName = process.env.SsoConnectionName;
        this.skillDialogId = skillDialog.id;

        this.addDialog(new ChoicePrompt(ACTION_STEP_PROMPT));
        this.addDialog(new SsoSignInDialog(this.connectionName));
        this.addDialog(skillDialog);

        this.addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
            this.promptActionStep.bind(this),
            this.handleActionStep.bind(this),
            this.promptFinalStep.bind(this)
        ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    async promptActionStep(stepContext) {
        const messageText = 'What SSO action do you want to perform?';
        const repromptMessageText = 'That was not a valid choice, please select a valid choice.';

        return await stepContext.prompt(DELIVERY_PROMPT, {
            prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
            retryPrompt: MessageFactory.text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
            choices: await this.getPromptChoices(stepContext)
        });
    }

    async getPromptChoices(stepContext) {
        let promptChoices = [];
        let adapter = stepContext.context.adapter;
        let token = await adapter.getUserToken(stepContext.context, this.connectionName);

        if (!token) {
            promptChoices.push('Login');
            // Token exchange will fail when the host is not logged on and the skill should
            // show a regular OAuthPrompt.
            promptChoices.push('Call Skill (without SSO)');
        } else {
            promptChoices.push('Logout');
            promptChoices.push('Show token');
            promptChoices.push('Call Skill (with SSO)');
        }

        promptChoices.push('Back');

        return ChoiceFactory.toChoices(promptChoices);
    }

    async handleActionStep(stepContext) {
        const action = stepContext.result.value;

        switch (action.toLowerCase()) {
            case 'login':
                return await stepContext.beginDialog(SSO_SIGNIN_DIALOG);
            case 'logout':
                const adapter = stepContext.context.adapter;
                await adapter.signOutUser(stepContext.context, this.connectionName);
                await stepContext.context.sendActivity('You have been signed out.');
                return await stepContext.next();
            case 'show token':
                const tokenProvider = stepContext.context.adapter;
                const token = await tokenProvider.getUserToken(stepContext.context, this.connectionName);
                if (!token) {
                    await stepContext.context.sendActivity('User has no cached token.');
                } else {
                    await stepContext.context.sendActivity(`Here is your current SSO token: ${ token.token }`);
                }
                return await stepContext.next();
            case 'call skill (with sso)':
            case 'call skill (without sso)':
                let beginSkillActivity = {
                    type: ActivityTypes.Event,
                    name: 'Sso'
                }
                return await stepContext.beginDialog(this.skillDialogId, { activity: beginSkillActivity });
            case 'back':
                await stepContext.context.sendActivity({
                    type: ActivityTypes.EndOfConversation,
                    code: EndOfConversationCodes.CompletedSuccessfully
                })
                return {status: DialogTurnStatus.complete};
            default:
                // This should never be hit since the previous prompt validates the choice.
                throw new Error(`[SsoDialog]: Unrecognized action: ${ action }`);
        }
    }

    async promptFinalStep(stepContext) {
        // Restart the dialog (we will exit when the user says end).
        return await stepContext.replaceDialog(this.initialDialogId);
    }
}

module.exports.SsoDialog = SsoDialog;
